Update Build Process of the Project and Improve Publishing

LukasBrand opened the pull request:
This Pull Request updates the project to use Gradle instead of Maven for a better overall build experience.

Main motivation behind this PR is the improved:
- publish pipeline to Maven Central
- artifact module metadata (module.json) for dependency resolution
- artifact metadata (Title, Vendor, Version) inside Jar
- maintenance of individual subprojects
- separation of dependencies and source sets
- reproducible artifacts
- CI by only using GitHub Actions
- javadoc generation
- easy future dependency update integration (e.g. Renovate)
- foundation work for future improvements (such as java module system, etc.)

The PR aims to NOT modify the created artifacts (Jars) beyond the Maven specific metadata and generally tries to keep the Java code untouched as much as possible (shared test code had to be moved so it can be consumed by other subprojects).
Configurations of build plugins were also migrated or adjusted to the project.

I kept the configuration of the individual subprojects rather separate. This can also be changed to shared configurations if it better reflects how the modules should be used.

The PR currently does not enforce formatting oder any checks such as spotbugs or checkstyle though it seemed as those tools were previously also not mandatory to run (tools find issues in existing code)

The PR gives a suggestion about how the artifacts could be automatically published in the future with the publish.yml GitHub Action workflow. On a GitHub release, this action is called and automatically published the artifacts to Maven Central. Currently, it publishes all the subprojects together which is similar to how it was probably handled before (every artifact of the repository is released at the same time). This can also be changed to be able to release them separately if desired.

The Publish GitHub Action requires the following Secrets to be set up:
```
ORG_GRADLE_PROJECT_signingKey: ${{ secrets.SIGNING_KEY }}
ORG_GRADLE_PROJECT_signingPassword: ${{ secrets.SIGNING_PASSWORD }}
ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.MAVEN_CENTRAL_USERNAME }}
ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.MAVEN_CENTRAL_PASSWORD }}
```
More details can be found here:
https://github.com/SgtSilvio/gradle-maven-central-publishing?tab=readme-ov-file

It might also be necessary to migrate the namespace away from Legacy OSSRH:
https://central.sonatype.org/faq/what-is-different-between-central-portal-and-legacy-ossrh/#self-service-migration

Thank you for creating this project!

LukasBrand committed:
Extract Gradle version catalog from Maven configuration

LukasBrand committed:
WIP: Add Gradle wrapper and root scaffold

LukasBrand committed:
WIP: Port microraft module to Gradle:
- Introduce test fixtures to share test code across project to other modules
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
WIP: Port microraft-hocon module to Gradle:
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)
- Update README

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
WIP: Port microraft-metrics module to Gradle:
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)
- Update README

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
WIP: Port microraft-store-sqlite module to Gradle:
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)
- Created README

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
WIP: Port microraft-tutorial module to Gradle:
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)
- Created README
- Removed test jar creation

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
WIP: Port microraft-yaml module to Gradle:
- Replicate current configuration from Maven
- Explicit in-module configuration
- Add additional information to Jar file (Title, Vendor, Version)
- Updated README

Missing:
- Publishing
- Staging
- Checksum

LukasBrand committed:
Add checkstyle plugin to root project. (It finds issues)

LukasBrand committed:
Add javadoc links plugin

LukasBrand committed:
Fix javadoc excludes

LukasBrand committed:
Add todos for root project

LukasBrand committed:
microraft: Add publishing and signing

LukasBrand committed:
microraft: Add dependency information to README

LukasBrand committed:
Remove maven from project

LukasBrand committed:
Update CI:
- Add job to release all artifacts
- Add job to test all artifacts
- Remove Circle CI

LukasBrand committed:
Move spotbugs and formatter files for now

LukasBrand committed:
Fix test shell file

LukasBrand committed:
Update root README:
- Update dependency links
- Update batches
- Update build instructions

LukasBrand committed:
- Add manual spotbugs and checkstyle
- Update documentation
- Ignore formatter for now
- Assume resources were declared accidentally (LICENSE and NOTICE)

LukasBrand committed:
Update build-site.sh

LukasBrand committed:
Update setup.md

SgtSilvio commented on gradle.properties:
```diff
@@ -0,0 +1,6 @@
+org.gradle.caching=true
+org.gradle.configuration-cache.parallel=true
+org.gradle.configuration-cache=true
+org.gradle.daemon.idletimeout=57600000
+org.gradle.jvmargs=-Xmx2g -XX:MaxMetaspaceSize=1g
```
```suggestion
```
probably too much copy&paste, the default memory settings should be enough

SgtSilvio commented on settings.gradle.kts:
```diff
@@ -0,0 +1,13 @@
+        gradlePluginPortal()
+    }
+}
+
```
```suggestion
```
this should be the default

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,85 @@
+description = ""
+
+metadata {
+    moduleName.set("io.microraft.microraft-hocon")
```
```suggestion
    moduleName = "io.microraft.microraft-hocon"
```
please use simple assignment in all build scripts

SgtSilvio replied:
```suggestion
    moduleName.set("io.microraft.hocon")
```
a module name should not just be a concatenation of group and artifact id, but a reverse domain name

LukasBrand committed:
Apply review comments

LukasBrand committed:
Add missing publishing to other subprojects

SgtSilvio commented on build.gradle.kts:
```diff
@@ -0,0 +1,9 @@
+}
+
+// Checkstyle and Spotbugs are applied for every project but do need to be triggered manually
+allprojects {
```
please apply the plugins in each project that requires it

LukasBrand committed:
use module declaration instead of group and name inside version catalog.

SgtSilvio commented on tools/repeat-test.sh:
```diff
@@ -13,16 +13,18 @@ fi
 
-mvn clean
+cd ..
+
```
```suggestion
```

SgtSilvio commented on README.md:
```diff
@@ -52,18 +56,33 @@ See [the User Guide](https://microraft.io/docs/setup).
-    <groupId>io.microraft</groupId>
-    <artifactId>microraft</artifactId>
-    <version>0.5</version>
+	<groupId>io.microraft</groupId>
```
```suggestion
    <groupId>io.microraft</groupId>
```
nitpick: tabs are not rendered nicely on GitHub

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+mockito = { module = "org.mockito:mockito-core", version.ref = "mockito" }
+slf4j = { module = "org.slf4j:slf4j-api", version.ref = "slf4j" }
+sqlite = { module = "org.xerial:sqlite-jdbc", version.ref = "sqlite" }
+typesafe = { module = "com.typesafe:config", version.ref = "typesafe" }
```
```suggestion
typesafe-config = { module = "com.typesafe:config", version.ref = "typesafe-config" }
```

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+mockito = "3.12.4"
+slf4j = "1.7.36"
+sqlite = "3.41.2.2"
+typesafe = "1.4.2"
```
```suggestion
typesafe-config = "1.4.2"
```

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+[libraries]
+assertj = { module = "org.assertj:assertj-core", version.ref = "assertj" }
+findbugs = { module = "com.google.code.findbugs:annotations", version.ref = "findbugs" }
+jackson = { module = "com.fasterxml.jackson.core:jackson-databind", version.ref = "jackson" }
```
```suggestion
jackson-databind = { module = "com.fasterxml.jackson.core:jackson-databind", version.ref = "jackson" }
```

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+log4j-slf4j-impl = { module = "org.apache.logging.log4j:log4j-slf4j-impl", version.ref = "log4j" }
+micrometer = { module = "io.micrometer:micrometer-core", version.ref = "micrometer" }
+mockito = { module = "org.mockito:mockito-core", version.ref = "mockito" }
+slf4j = { module = "org.slf4j:slf4j-api", version.ref = "slf4j" }
```
```suggestion
slf4j-api = { module = "org.slf4j:slf4j-api", version.ref = "slf4j" }
```

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+slf4j = { module = "org.slf4j:slf4j-api", version.ref = "slf4j" }
+sqlite = { module = "org.xerial:sqlite-jdbc", version.ref = "sqlite" }
+typesafe = { module = "com.typesafe:config", version.ref = "typesafe" }
+yaml = { module = "org.yaml:snakeyaml", version.ref = "yaml" }
```
```suggestion
snakeyaml = { module = "org.yaml:snakeyaml", version.ref = "snakeyaml" }
```

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+slf4j = "1.7.36"
+sqlite = "3.41.2.2"
+typesafe = "1.4.2"
+yaml = "2.0"
```
```suggestion
snakeyaml = "2.0"
```

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,104 @@
+}
+
+tasks.javadoc {
+    isFailOnError = false
```
why is this config required? are there javadoc errors?

LukasBrand replied:
Yes, I applied the same settings as before. I did not want to touch the code to be able to remove this.

LukasBrand replied:
I originally missed some exclusions which I added later on. That seems to have fixed the errors. Removed the flag now. Good catch!

SgtSilvio commented on pom.xml:
```diff
@@ -1,455 +0,0 @@
-            <resource>
-                <directory>${root.dir}</directory>
-                <includes>
-                    <include>NOTICE*</include>
-                    <include>LICENSE*</include>
```
These files are now missing from the jar files or did I miss the config?

SgtSilvio replied:
Looking at a previous release, it seems that these files were actually not part of the jar file.

LukasBrand replied:
Exactly! Hence it was never inside the jar I decided to also not add it now. 

I assume the code inside the pom was ignored because it would only apply to the jar of the root project which has none.

SgtSilvio commented on microraft-metrics/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+}
+
+tasks.javadoc {
+    isFailOnError = false
```
same as above: why is this line configured?

SgtSilvio commented on microraft-tutorial/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+    alias(libs.plugins.javadocLinks)
+    `maven-publish`
+    signing
+    alias(libs.plugins.mavenCentralPublishing)
```
does publishing of the tutorial on maven central make sense?

LukasBrand replied:
It was published previously, I kept it as-is.

SgtSilvio commented on microraft-metrics/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+}
+
+dependencies {
+    implementation(project(":microraft"))
+    implementation(libs.micrometer)
```
```suggestion
    api(project(":microraft"))
    api(libs.micrometer)
```

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,104 @@
+}
+
+dependencies {
+    implementation(project(":microraft"))
+    implementation(libs.typesafe)
```
```suggestion
    api(project(":microraft"))
    api(libs.typesafe)
```

SgtSilvio commented on microraft-store-sqlite/build.gradle.kts:
```diff
@@ -0,0 +1,107 @@
+}
+
+dependencies {
+    implementation(project(":microraft"))
```
```suggestion
    api(project(":microraft"))
```

SgtSilvio commented on microraft-yaml/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+
+dependencies {
+    implementation(project(":microraft"))
+    implementation(libs.yaml)
```
```suggestion
    api(project(":microraft"))
    api(libs.yaml)
```

SgtSilvio commented on microraft/build.gradle.kts:
```diff
@@ -0,0 +1,132 @@
+            dependencies {
+                implementation(libs.mockito)
+                implementation(libs.assertj)
+                implementation(libs.log4j.core)
+                implementation(libs.log4j.api)
+                implementation(libs.log4j.slf4j.impl)
```
these log4j dependencies can probably be runtimeOnly
log4j-api can then also be removed if it is not used during compile time

SgtSilvio commented on microraft/build.gradle.kts:
```diff
@@ -0,0 +1,132 @@
+
+
+publishing {
+    // TODO Remove after debugging
```
I guess this should now be removed

LukasBrand replied:
I meant to leave it in until signing etc. is figured out.

SgtSilvio commented on pom.xml:
```diff
@@ -1,455 +0,0 @@
-        <assertj.version>3.23.1</assertj.version>
-        <findbugs.annotations.version>3.0.1</findbugs.annotations.version>
-
-        <surefire.fork.count>4</surefire.fork.count>
```
vmHeapSettings and surefire.fork.count are not configured for the test tasks (maybe they are unnecessary but at least the test execution times should be checked if they are now longer)

SgtSilvio commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,37 @@
+
+[libraries]
+assertj = { module = "org.assertj:assertj-core", version.ref = "assertj" }
+findbugs = { module = "com.google.code.findbugs:annotations", version.ref = "findbugs" }
```
```suggestion
findbugs-annotations = { module = "com.google.code.findbugs:annotations", version.ref = "findbugs" }
```

metanet commented:
fantastic work, folks! let me know when the PR is finalized, then i can review and merge

LukasBrand committed:
Apply review:
- Remove change directory for test file
- Format READMEs
- Improve version catalog naming
- Remove unnecessary javadoc flag (it seems like the exclusion fixed it)
- use api configuration for some transitive dependencies

LukasBrand committed:
Suppress unused test JvmTestSuite

LukasBrand committed:
Cleanup log4j inside microraft:test

LukasBrand committed:
Configure parallel tests for the microraft project

LukasBrand committed:
Move spotbugs and checkstyle into projects

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+    )
+}
+
+
+repositories {
+    mavenCentral()
+}
+
```
```suggestion
```
We already have centralized dependency declarations in libs.versions.toml, let's also centralize the repository declarations (in this project it is anyhow only mavenCentral). To do so, add the following to settings.gradle.kts:
```
dependencyResolutionManagement {
    @Suppress("UnstableApiUsage") //
    repositories {
        mavenCentral()
    }
}
```

SgtSilvio commented on microraft/build.gradle.kts:
```diff
@@ -0,0 +1,138 @@
+            dependencies {
+                implementation(libs.mockito)
+                implementation(libs.assertj)
+                runtimeOnly(libs.log4j.core)
```
```suggestion
```
`libs.log4j.slf4j.impl` brings all necessary dependencies

LukasBrand replied:
Agreed on `org.apache.logging.log4j:log4j-slf4j-impl` can also be considered as a log4j impl, so we do not need a separate one.

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        @Suppress("unused") //
+        val test by getting(JvmTestSuite::class) {
```
```suggestion
        named<JvmTestSuite>("test") {
```

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+@Suppress("UnstableApiUsage") //
+testing {
+    suites {
+        withType(JvmTestSuite::class) {
```
```suggestion
        withType<JvmTestSuite> {
```

SgtSilvio commented on microraft-store-sqlite/build.gradle.kts:
```diff
@@ -0,0 +1,106 @@
+                implementation(testFixtures(project(":microraft")))
+                implementation(libs.assertj)
+                implementation(libs.jackson.databind)
+                compileOnly(libs.findbugs.annotations)
```
```suggestion
                runtimeOnly(libs.log4j.slf4j.impl)
                compileOnly(libs.findbugs.annotations)
```

SgtSilvio commented on microraft-tutorial/build.gradle.kts:
```diff
@@ -0,0 +1,102 @@
+        @Suppress("unused") //
+        val test by getting(JvmTestSuite::class) {
+            dependencies {
+                implementation(libs.assertj)
```
```suggestion
                implementation(libs.assertj)
                runtimeOnly(libs.log4j.slf4j.impl)
```

LukasBrand committed:
Centralize repositories

LukasBrand committed:
Remove unnecessary log4j-core

SgtSilvio commented on microraft-hocon/build.gradle.kts:
```diff
@@ -0,0 +1,103 @@
+
+group = "io.microraft"
+version = "0.9-SNAPSHOT"
+description = ""
```
we probably should add proper descriptions

SgtSilvio replied:
suggestion:
```suggestion
description = "HOCON config parser for MicroRaft"
```

LukasBrand committed:
Apply review:
- Minor test api usage change
- Add logging dependency to test source sets where necessary

LukasBrand committed:
Minor documentation fix

SgtSilvio commented on microraft-metrics/build.gradle.kts:
```diff
@@ -0,0 +1,102 @@
+
+group = "io.microraft"
+version = "0.9-SNAPSHOT"
+description = ""
```
suggestion:
```suggestion
description = "Micrometer metrics for MicroRaft"
```

SgtSilvio commented on microraft-store-sqlite/build.gradle.kts:
```diff
@@ -0,0 +1,106 @@
+
+group = "io.microraft"
+version = "0.9-SNAPSHOT"
+description = ""
```
suggestion:
```suggestion
description = "SQLite store for MicroRaft"
```

SgtSilvio commented on microraft-yaml/build.gradle.kts:
```diff
@@ -0,0 +1,97 @@
+
+group = "io.microraft"
+version = "0.9-SNAPSHOT"
+description = ""
```
suggestion:
```suggestion
description = "YAML config parser for MicroRaft"
```

SgtSilvio commented on microraft-tutorial/build.gradle.kts:
```diff
@@ -0,0 +1,98 @@
+
+group = "io.microraft"
+version = "0.9-SNAPSHOT"
+description = ""
```
```suggestion
description = "Tutorial for MicroRaft"
```

SgtSilvio commented on microraft-tutorial/build.gradle.kts:
```diff
@@ -0,0 +1,98 @@
+dependencies {
+    implementation(project(":microraft"))
+    implementation(libs.slf4j.api)
+    runtimeOnly(libs.log4j.slf4j.impl)
```
I think the logging impl was now unintentionally added to the main dependencies, it should have been added to the test dependencies

SgtSilvio replied:
@LukasBrand this is done, please resolve

SgtSilvio commented on microraft-tutorial/build.gradle.kts:
```diff
@@ -0,0 +1,98 @@
+
+
+dependencies {
+    implementation(project(":microraft"))
```
```suggestion
    api(project(":microraft"))
```

SgtSilvio replied:
@LukasBrand this is done, please resolve

SgtSilvio committed:
Move logging dependency to test dependencies

SgtSilvio committed:
Add project descriptions

SgtSilvio committed:
Cleanup empty lines

SgtSilvio committed:
Cleanup dependency order

SgtSilvio committed:
Move dependencyResolutionManagement under subproject includes

SgtSilvio commented:
@metanet this PR is now ready for review

SgtSilvio committed:
Run only tests (not checkstyle and spotbugs) in check github workflow

LukasBrand commented:
Hi @metanet 👋, 
please let me know if there's anything I can do to help with the review or make the process easier.

metanet commented:
sorry for the delay. i was abroad last week. thanks a lot for the extensive work here.
